File: services/payment-service/src/models/payment.model.ts
import { MAX_SETTLE_ATTEMPTS, type PaymentStatus, type SimulateMode } from '../domain/payment';

export interface PaymentRow {
  id: string;
  from_account_id: string;
  to_account_id: string;
  amount_cents: number;
  note: string | null;
  status: PaymentStatus;
  failure_reason: string | null;
  simulate_mode: SimulateMode;
  attempts: number;
  next_attempt_at: Date;
  correlation_id: string | null;
  created_at: Date;
  updated_at: Date;
}

export interface PaymentDto {
  id: string;
  fromAccountId: string;
  toAccountId: string;
  amountCents: number;
  note: string | null;
  status: PaymentStatus;
  failureReason: string | null;
  simulateMode: SimulateMode;
  attempts: number;
  maxAttempts: number;
  nextAttemptAt: Date;
  createdAt: Date;
  updatedAt: Date;
}

export const toPaymentDto = (row: PaymentRow): PaymentDto => ({
  id: row.id,
  fromAccountId: row.from_account_id,
  toAccountId: row.to_account_id,
  amountCents: row.amount_cents,
  note: row.note,
  status: row.status,
  failureReason: row.failure_reason,
  simulateMode: row.simulate_mode,
  attempts: row.attempts,
  // Sent alongside the count so a client can render "attempt 2 of 3" without
  // hardcoding the policy.
  maxAttempts: MAX_SETTLE_ATTEMPTS,
  nextAttemptAt: row.next_attempt_at,
  createdAt: row.created_at,
  updatedAt: row.updated_at,
});

/** The event body every payment lifecycle event shares. */
export const toEventBody = (row: PaymentRow, occurredAt: Date) => ({
  paymentId: row.id,
  fromAccountId: row.from_account_id,
  toAccountId: row.to_account_id,
  amountCents: row.amount_cents,
  note: row.note,
  occurredAt: occurredAt.toISOString(),
});

File: services/payment-service/src/main/java/com/ledgerflow/payment/models/PaymentModel.java
package com.ledgerflow.payment.models;

import com.ledgerflow.payment.domain.PaymentStatus;
import com.ledgerflow.payment.domain.Payments;
import com.ledgerflow.payment.domain.SimulateMode;
import com.ledgerflow.payment.lib.Iso;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class PaymentModel {

  private PaymentModel() {}

  /** @param holdReasons why the risk screen held this payment. Empty otherwise. */
  public record PaymentRow(
      String id,
      String fromAccountId,
      String toAccountId,
      long amountCents,
      String note,
      PaymentStatus status,
      String failureReason,
      SimulateMode simulateMode,
      int attempts,
      Instant nextAttemptAt,
      List<String> holdReasons,
      String correlationId,
      Instant createdAt,
      Instant updatedAt) {}

  public record PaymentDto(
      String id,
      String fromAccountId,
      String toAccountId,
      long amountCents,
      String note,
      PaymentStatus status,
      String failureReason,
      SimulateMode simulateMode,
      int attempts,
      int maxAttempts,
      Instant nextAttemptAt,
      List<String> holdReasons,
      Instant createdAt,
      Instant updatedAt) {}

  public static PaymentDto toPaymentDto(PaymentRow row) {
    return new PaymentDto(
        row.id(),
        row.fromAccountId(),
        row.toAccountId(),
        row.amountCents(),
        row.note(),
        row.status(),
        row.failureReason(),
        row.simulateMode(),
        row.attempts(),
        // Sent alongside the count so a client can render "attempt 2 of 3" without
        // hardcoding the policy.
        Payments.MAX_SETTLE_ATTEMPTS,
        row.nextAttemptAt(),
        row.holdReasons() == null ? List.of() : row.holdReasons(),
        row.createdAt(),
        row.updatedAt());
  }

  /** The event body every payment lifecycle event shares. */
  public static Map<String, Object> toEventBody(PaymentRow row, Instant occurredAt) {
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("paymentId", row.id());
    body.put("fromAccountId", row.fromAccountId());
    body.put("toAccountId", row.toAccountId());
    body.put("amountCents", row.amountCents());
    body.put("note", row.note());
    body.put("occurredAt", Iso.format(occurredAt));
    return body;
  }
}
